src/snowflake/cli/plugins/nativeapp/run_processor.py

```diff
@@ -311,24 +276,25 @@ def upgrade_app(
             # 3. If existing application is found, perform a few validations and upgrade the application object.
             if show_app_row:
 
-                # We skip comment check here, because prod/pre-existing application objects may not be created by the Snowflake CLI.
-                # Check for the right owner
-                ensure_correct_owner(
-                    row=show_app_row, role=self.app_role, obj_name=self.app_name
-                )
+                install_method.ensure_app_usable(self._na_project, show_app_row)
 
                 # If all the above checks are in order, proceed to upgrade
                 try:
+                    cc.step(f"Upgrading existing application object {self.app_name}.")
+                    using_clause = install_method.using_clause(self._na_project)
                     self._execute_query(
                         f"alter application {self.app_name} upgrade {using_clause}"
                     )
 
-                    if using_clause:
+                    if install_method.is_dev_mode:
                         # if debug_mode is present (controlled), ensure it is up-to-date
                         if self.debug_mode is not None:
                             self._execute_query(
                                 f"alter application {self.app_name} set debug_mode = {self.debug_mode}"
                             )
+
+                    # hooks always executed after a create or upgrade
+                    self._execute_post_deploy_hooks()
```

Comment:
@sfc-gh-gbloom post-deploy hooks weren't being executed on upgrade -- intentional?

Reply:
Are you sure they weren't? On upgrade they should be executed [here](https://github.com/snowflakedb/snowflake-cli/blob/main/src/snowflake/cli/plugins/nativeapp/run_processor.py#L117), and the flow should be covered by [this test](https://github.com/snowflakedb/snowflake-cli/blob/main/tests_integration/nativeapp/test_init_run.py#L448-L465).

Reply:
That function only applied to dev mode apps. The previous `upgrade_app` method was used for versioned dev and from-release-directive modes.

Reply:
Would have expected it to be right before this return statement: https://github.com/snowflakedb/snowflake-cli/blob/main/src/snowflake/cli/plugins/nativeapp/run_processor.py#L332

Reply:
Good catch, these flows were not covered

Reply:
I will keep parity in this PR and fix in a subsequent one.